rbd: remove topologyConstrainedPools parameter

iPraveenParihar opened the pull request:
<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This commit removes the `topologyConstrainedPools` parameter from PV volumeAttributes as it is not required.

## Related issues ##

Fixes: #4497 

## Future concerns ##

List items that are not part of the PR and do not impact it's
functionality, but are work items that can be taken up subsequently.

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [x] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [x] Documentation has been updated, if necessary.
* [x] Unit tests have been added, if necessary.
* [x] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

iPraveenParihar committed:
rbd: remove topologyConstrainedPools parameter

This commit removes the `topologyConstrainedPools` parameter
from PV volumeAttributes as it is not required.

Signed-off-by: Praveen M <m.praveen@ibm.com>

iPraveenParihar commented:
verified that `topologyConstrainedPools` has been removed. 

```shell
[pm@dhcp53-176 examples]$ k get pv pvc-5aaf7aef-fda5-4a57-af7c-abbb737d77bb -o jsonpath='{.spec.csi.volumeAttributes}' | jq
{
  "clusterID": "rook-ceph",
  "encryptionKMSID": "azure-kms",
  "encryptionType": "block",
  "imageFeatures": "layering",
  "imageName": "csi-vol-1292393c-b86a-4ff7-abb7-52750879d83a",
  "journalPool": "replicapool",
  "pool": "rp",
  "storage.kubernetes.io/csiProvisionerIdentity": "1710418563048-5228-rook-ceph.rbd.csi.ceph.com"
}
```

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -252,6 +252,9 @@ func buildCreateVolumeResponse(req *csi.CreateVolumeRequest, rbdVol *rbdVolume)
 		}
 	}
 
+	// remove topologyConstrainedPools param as it is not required
```
It feels cleaner to do this at the place where other keys to `volumeContext` are added or removed. Maybe directly below `k8s.RemoveCSIPrefixedParameters()` ?

iPraveenParihar replied:
yea, that will be good.
How about introducing a function called `RemoveVolumeContextParameters`? This function could serve the specific purpose of removing parameters from the volume context."

nixpanic replied:
Sure, or make that a `volumeContext := rbdVol.getVolumeContext(req.GetParameters())` function.

nixpanic commented on internal/rbd/controllerserver.go:
```diff
@@ -225,8 +225,10 @@ func (cs *ControllerServer) parseVolCreateRequest(
+	
 	// remove kubernetes csi prefixed parameters.
-	volumeContext := k8s.RemoveCSIPrefixedParameters(req.GetParameters())
+	volumeContext = k8s.RemoveCSIPrefixedParameters(req.GetParameters())
```
Move `k8s.RemoveCSIPrefixedParameters()` into your new `rbdVol.getVolumeContext()` function

Madhu-1 commented on internal/rbd/rbd_util.go:
```diff
@@ -2155,3 +2156,24 @@ func (rv *rbdVolume) unsetAllMetadata(keys []string) error {
 }
+
+// getVolumeContext filters out parameters that are not required in the volume context.
+func (rv *rbdVolume) getVolumeContext(parameters map[string]string) map[string]string {
```
we dont need to make this as a method, moving this to outside util is good and we can pass the notRequiredParams so that we can reuse this same function for cephfs as well.

iPraveenParihar replied:
@Madhu-1 
will move this to `internal/util/util.go`  and rename as `RemoveParameters(params, notRequiredParams)` ?

Madhu-1 replied:
Yes that also helps

iPraveenParihar committed:
cephfs: refactor code for improved reusability

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented on internal/util/topology.go:
```diff
@@ -30,6 +30,9 @@ import (
 	labelSeparator string = ","
+
+	// topologyPoolsParam is the parameter name used to pass topology constrained pools.
+	TopologyPoolsParam = "topologyConstrainedPools"
```
this is not use outside of the package, so you can call it `topologyPoolsParam` as in the comment

nixpanic commented on internal/util/util.go:
```diff
@@ -390,3 +391,34 @@ func CallStack() string {
 }
+
+// CheckSliceContains checks the slice for string.
+func CheckSliceContains(options []string, opt string) bool {
```
`k8s.io/utils/strings/slices` is already used elsewhere, can you replace `CheckSliceContains()` by `slices.Contains()`?

nixpanic replied:
Wait, no! There is https://pkg.go.dev/slices in recent Golang versions :partying_face: , use that instead.

iPraveenParihar replied:
done😃

iPraveenParihar committed:
cleanup: use slices package

This commit replaces the user implemented function
`CheckSliceContains()` with `slices.Contains()`
function introduced in Go 1.21.

Signed-off-by: Praveen M <m.praveen@ibm.com>

nixpanic commented:
@mergifyio rebase

nixpanic commented:
@mergifyio queue

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.29

nixpanic commented:
> /retest ci/centos/mini-e2e/k8s-1.29

spurious failure, the [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.29/detail/mini-e2e_k8s-1.29/165/pipeline) show that an image was not cleaned up (yet?):
```
[FAILED] backend images not matching kubernetes resource count,image count 1 kubernetes resource count 0
```

nixpanic commented:
@mergifyio requeue

Madhu-1 commented:
> > /retest ci/centos/mini-e2e/k8s-1.29
> 
> spurious failure, the [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.29/detail/mini-e2e_k8s-1.29/165/pipeline) show that an image was not cleaned up (yet?):
> 
> ```
> [FAILED] backend images not matching kubernetes resource count,image count 1 kubernetes resource count 0
> ```

i think we need to update the validation function to print `rbd info` and `rbd status` of each images to get more clarity what happened. it looks like the image is opened by someone but not closed? not sure who

nixpanic commented:
@mergifyio queue
